### ISSUE TITLE
no-unused-expressions makes an error for optional chaining of method calls: foo?.bar();

### ISSUE THREAD

**Issue by Frozen-byte** [eslint/eslint#12822]
Reviving #11045 since optional chaining arrived Stage4!

I am unable to write a heads-up to the Issue itself. 
This conversation has been locked and limited to collaborators.

**Comment by kaicataldo** [comment]
We're currently blocked by https://github.com/estree/estree/pull/204. Once that lands, all of the core rules will be updated to handle this syntax. Thanks for your patience!

**Comment by tom-sherman** [comment]
Slightly adjacent, what is the reason for the "no-unused-expressions" rule being matched for statements that cannot be parsed?

**Comment by kaicataldo** [comment]
My guess is that the issue author is using a custom parser that can parse this. When using the default parser, this would currently be a parser error.

**Comment by Frozen-byte** [comment]
Yes I am using "babel/no-unused-expressions" to get around that other issue.

**Comment by roth-j** [comment]
Has this issue been addressed? I am still seeing the eslint error for an unused expression when using optional chaining
```
class My Component {

...

  onChange = evt => {
    this.props.onChange?.(evt.target.value); // throws no-unused-expression
  }; 

  noChangeAlt = evt => {
    this.props.onChange(evt.target.value); // no eslint error but its complaining about executing a possibly undefined thing
  }

**Comment by garrett-gottlieb** [comment]
In the meantime, you can ignore with `"no-unused-expressions": "off"`.

**Comment by kaicataldo** [comment]
As [mentioned above](https://github.com/eslint/eslint/issues/12822#issuecomment-577475930), we're still blocked by ESTree deciding on the AST spec.

**Comment by tomasklingen** [comment]
estree/estree#204 is merged! :)

**Comment by jeiea** [comment]
Any update on this?

**Comment by mdjermanovic** [comment]
The core `no-unused-expressions` rule will support optional chaining per the [`estree`](https://github.com/estree/estree/blob/master/es2020.md#chainexpression) specification when we release #13416 (expected to be in v7.5.0).

At that point, the rule will work well with the default parser.

However, the current version of `babel-eslint` parser generates a different AST so the rule will probably not work well until  `babel-eslint` releases babel/babel#11815.

Until the new `babel-eslint` release, you should probably use [`babel-eslint-plugin`](https://github.com/babel/eslint-plugin-babel) and [`babel/no-unused-expressions`](https://github.com/babel/eslint-plugin-babel#rules) rule instead of the core `no-unused-expressions` rule.

**Comment by jeiea** [comment]
#13416 and babel/babel#11815 seems to be released. I had no issue without babel-eslint-plugin. Can we close this issue?

**Comment by mdjermanovic** [comment]
Yes, I think we can close this issue.

We also changed the rule to not warn on unknown expressions when used with third-party parsers.

I can confirm now that the core `no-unused-expressions` rule doesn't report an error on `foo?.bar();` when used with any of the following parsers: `espree` (default parser), `babel-eslint`, `@babel/eslint-parser`, `@typescript-eslint/parser`.
